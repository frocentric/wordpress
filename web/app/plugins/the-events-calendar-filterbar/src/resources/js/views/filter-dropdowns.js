/* global tribe, tribe_dropdowns */
/* eslint-disable no-var, strict */

/**
 * Makes sure we have all the required levels on the Tribe Object.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar = tribe.filterBar || {};

/**
 * Configures Filter Dropdowns Object in the Global Tribe variable.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar.filterDropdowns = {};

/**
 * Initializes in a Strict env the code that manages the filter dropdowns.
 *
 * @since 5.0.0
 *
 * @param  {PlainObject} $   jQuery
 * @param  {PlainObject} obj tribe.filterBar.filterDropdowns
 *
 * @return {void}
 */
( function( $, obj ) {
	'use strict';
	var $document = $( document );

	/**
	 * Selectors used for configuration and setup.
	 *
	 * @since 5.0.0
	 *
	 * @type {PlainObject}
	 */
	obj.selectors = {
		dropdownInput: '[data-js="tribe-filter-bar-c-dropdown-input"]',
	};

	/**
	 * Handle dropdown change event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object of change event.
	 *
	 * @return {void}
	 */
	obj.handleDropdownChange = function( event ) {
		var key = event.data.target.attr( 'name' );

		// Return early if name attribute is not set.
		if ( ! key ) {
			return;
		}

		// Return early if an ajax request is already happening.
		if ( tribe.events.views.manager.currentAjaxRequest ) {
			return;
		}

		var location = tribe.filterBar.filters.removeKeyValueFromQuery( window.location, key, true );
		var value = event.data.target.attr( 'value' );

		if ( value ) {
			location = tribe.filterBar.filters.addKeyValueToQuery( location, key, value );
		}

		tribe.filterBar.filters.submitRequest( event.data.container, location.href );
	};

	/**
	 * Handler for template selection.
	 *
	 * @since  5.0.0
	 *
	 * @param  {PlainObject} state State of selected item.
	 *
	 * @return {jQuery}
	 */
	obj.handleTemplateSelection = function( state ) {
		var $newEl = $( '<span class="select2-selection__choice__text"></span>' );
		$newEl.text( state.text );
		return $newEl;
	};

	/**
	 * Initializes filter dropdown.
	 *
	 * @param  {jQuery}  $dropdownInput jQuery object of dropdown input.
	 * @param  {jQuery}  $container     jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.initDropdown = function( $dropdownInput, $container ) {
		tribe_dropdowns.dropdown( $dropdownInput, {
			templateSelection: obj.handleTemplateSelection,
		} );
		$dropdownInput
			.on( 'change', { target: $dropdownInput, container: $container }, obj.handleDropdownChange )
			.data( 'select2' ).$container.addClass( 'select2-container--open' );
		$dropdownInput.data( 'select2' ).trigger( 'query', {} );
	};

	/**
	 * Deinitializes filter dropdowns.
	 *
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.deinitDropdowns = function( $container ) {
		$container
			.find( obj.selectors.dropdownInput )
			.each( function( index, dropdownInput ) {
				$( dropdownInput )
					.off()
					.select2( 'destroy' );
			} );
	};

	/**
	 * Initializes filter dropdowns.
	 *
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.initDropdowns = function( $container ) {
		$container
			.find( obj.selectors.dropdownInput )
			.each( function( index, dropdownInput ) {
				obj.initDropdown( $( dropdownInput ), $container );
			} );
	};

	/**
	 * Deinitialize filter dropdowns JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object for 'beforeAjaxSuccess.tribeEvents' event.
	 *
	 * @return {void}
	 */
	obj.deinit = function( event ) {
		var $container = event.data.container;
		obj.deinitDropdowns( $container );
		$container.off( 'beforeAjaxSuccess.tribeEvents', obj.deinit );
	};

	/**
	 * Initialize filter dropdowns JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event}   event      event object for 'afterSetup.tribeEvents' event.
	 * @param  {integer} index      jQuery.each index param from 'afterSetup.tribeEvents' event.
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.init = function( event, index, $container ) {
		obj.initDropdowns( $container );
		$container.on( 'beforeAjaxSuccess.tribeEvents', { container: $container }, obj.deinit );
	};

	/**
	 * Handles the initialization of filter dropdowns when Document is ready.
	 *
	 * @since 5.0.0
	 *
	 * @return {void}
	 */
	obj.ready = function() {
		$document.on( 'afterSetup.tribeEvents', tribe.events.views.manager.selectors.container, obj.init );
	};

	// Configure on document ready.
	$( obj.ready );
} )( jQuery, tribe.filterBar.filterDropdowns );
