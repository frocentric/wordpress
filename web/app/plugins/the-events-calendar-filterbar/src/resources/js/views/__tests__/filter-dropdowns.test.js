describe( 'Filter Dropdowns', () => {
	beforeAll( () => {
		String.prototype.className = function() {
			if (
				(
					'string' !== typeof this &&
					! this instanceof String /* eslint-disable-line no-unsafe-negation */
				) ||
				'function' !== typeof this.replace
			) {
				return this;
			}

			return this.replace( '.', '' );
		};

		global.tribe = {};
		require( '../filter-dropdowns' );
		global.tribe.filterBar.filters = {};
		global.tribe_dropdowns = {};
	} );

	afterAll( () => {
		delete String.prototype.className;
		delete global.tribe;
		delete global.tribe_dropdowns;
	} );

	describe( 'Selectors', () => {
		test( 'Should match snapshot', () => {
			const selectors = JSON.stringify( tribe.filterBar.filterDropdowns.selectors );
			expect( selectors ).toMatchSnapshot();
		} );
	} );

	describe( 'Handle dropdown change', () => {
		beforeEach( () => {
			global.tribe.events = {
				views: {
					manager: {
						currentAjaxRequest: null,
					},
				},
			};
			global.tribe.filterBar.filters = {
				addKeyValueToQuery: jest.fn().mockImplementation( () => ( {} ) ),
				removeKeyValueFromQuery: jest.fn().mockImplementation( () => ( {} ) ),
				submitRequest: jest.fn(),
			};
		} );

		test( 'Should return early if name attribute is empty', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterDropdowns.handleDropdownChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );
		} );

		test( 'Should return early if an ajax request is already happening', () => {
			// Setup test.
			global.tribe.events = {
				views: {
					manager: {
						currentAjaxRequest: {},
					},
				},
			};
			const event = {
				data: {
					target: $( '<div name="key" value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterDropdowns.handleDropdownChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );

			// Cleanup test.
			delete global.tribe.events.views.manager;
		} );

		test( 'Should remove key value from query', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div name="key"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterDropdowns.handleDropdownChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should replace key value in query with selected option', () => {
			// Setup test.
			const event = {
				data: {
					target: $( '<div name="key" value="42"></div>' ),
					container: $( '<div></div>' ),
				},
			};

			// Test.
			tribe.filterBar.filterDropdowns.handleDropdownChange( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.addKeyValueToQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Handle template selection', () => {
		test( 'Should return selected option wrapped in span element', () => {
			// Setup test.
			const state = {
				text: 'Hello world',
			};

			// Test.
			const $template = tribe.filterBar.filterDropdowns.handleTemplateSelection( state );

			// Confirm final states.
			expect( $template ).toMatchSnapshot();
		} );
	} );

	describe( 'Initialize dropdown', () => {
		test( 'Should initialize dropdown', () => {
			// Setup test.
			global.tribe_dropdowns = {
				dropdown: jest.fn(),
			};
			const $dropdownInput = $( '<input />' );
			const $container = $( '<div></div>' );
			$dropdownInput.on = jest.fn().mockImplementation( () => $dropdownInput );
			const addClass = jest.fn();
			const trigger = jest.fn();
			$dropdownInput.data = () => ( {
				$container: {
					addClass: addClass,
				},
				trigger: trigger,
			} );

			// Test.
			tribe.filterBar.filterDropdowns.initDropdown( $dropdownInput, $container );

			// Confirm final states.
			expect( global.tribe_dropdowns.dropdown.mock.calls.length ).toBe( 1 );
			expect( $dropdownInput.on.mock.calls.length ).toBe( 1 );
			expect( addClass.mock.calls.length ).toBe( 1 );
			expect( trigger.mock.calls.length ).toBe( 1 );
			expect( $dropdownInput.on.mock.calls[ 0 ][ 0 ] ).toBe( 'change' );
		} );
	} );
} );
