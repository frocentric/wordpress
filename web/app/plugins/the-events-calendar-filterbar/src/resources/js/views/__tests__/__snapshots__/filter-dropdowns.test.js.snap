// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Dropdowns Handle template selection Should return selected option wrapped in span element 1`] = `
jQuery {
  "0": <span
    class="select2-selection__choice__text"
  >
    Hello world
  </span>,
  "length": 1,
}
`;

exports[`Filter Dropdowns Selectors Should match snapshot 1`] = `"{\\"dropdownInput\\":\\"[data-js=\\\\\\"tribe-filter-bar-c-dropdown-input\\\\\\"]\\"}"`;
